### PR TITLE
Add epoch store to SafeClient

### PR DESCRIPTION
**Pull request by lxfind**
This PR adds `Arc<EpochStore>` to SafeClient. This allows us to read the corresponding committee from the right epoch, when we try to verify an authenticated data structure (based on the epoch from the signature).
A large part of this PR is due to the refactoring needed in order to be able to pass through the epoch store into SafeClient through AuthorityAggregator.

### CODE REVIEW

File: crates/sui-core/src/authority.rs [hunk 1]
```diff
@@ -344,6 +344,10 @@ impl AuthorityState {
         self.committee.load().epoch
     }
 
+    pub fn clone_epoch_store(&self) -> Arc<EpochStore> {
```
Comment (gdanezis) [inline crates/sui-core/src/authority.rs:347]: Returning a `&Arc` is my usual way of doing this to let the called decide what they need.

File: crates/sui-core/src/authority.rs [hunk 2]
```diff
@@ -1045,18 +1048,11 @@ impl AuthorityState {
...
-                .expect("Init genesis epoch data must not fail");
-            genesis_committee
-        } else {
-            epoch_store
-                .get_latest_authenticated_epoch()
-                .epoch_info()
-                .committee()
-                .clone()
-        };
+        let committee = epoch_store
```
Comment (gdanezis) [inline crates/sui-core/src/authority.rs:1051]: Here we get the very last committee, which I think is the one compatible with the latest written stuff in the DB when we start. However, this may not be the actual latest committee of the node has been away for a while. Do we pass that in a different place, or do we only need this when we sync?

File: crates/sui-core/src/epoch/epoch_store.rs [hunk 3]
```diff
@@ -19,12 +20,21 @@ pub struct EpochStore {
 }
 
 impl EpochStore {
-    pub fn new(path: PathBuf) -> Self {
-        Self::open_tables_read_write(path, None)
+    pub fn new(path: PathBuf, genesis_committee: &Committee) -> Self {
+        let epoch_store = Self::open_tables_read_write(path, None);
```
Comment (gdanezis) [inline crates/sui-core/src/epoch/epoch_store.rs:24]: Is this None DB options? We should always allow callers to define DB options, so lets pass in an Option.

File: crates/sui-core/src/epoch/tests/reconfiguration_tests.rs [hunk 4]
```diff
@@ -42,7 +42,7 @@ async fn test_start_epoch_change() {
     let state = states[0].clone();
 
     // Check that we initialized the genesis epoch.
-    let init_epoch = state.epoch_store.get_latest_authenticated_epoch();
+    let init_epoch = state.clone_epoch_store().get_latest_authenticated_epoch();
```
Comment (gdanezis) [inline crates/sui-core/src/epoch/tests/reconfiguration_tests.rs:45]: As per above: here I think there is no need to clone the Arc inside.

File: crates/sui-core/src/safe_client.rs [hunk 5]
```diff
@@ -56,7 +68,8 @@ impl<C> SafeClient<C> {
     ) -> SuiResult {
         if let Some(signed_transaction) = &response.signed_transaction {
             // Check the transaction signature
-            signed_transaction.verify(&self.committee)?;
+            signed_transaction
+                .verify(&self.get_committee(&signed_transaction.auth_sign_info.epoch)?)?;
```
Comment (gdanezis) [inline crates/sui-core/src/safe_client.rs:72]: I think the committee should be the same throughout the call, so why not take it once, and re-use it to make this clear? (And to avoid it changing half way through).

File: crates/sui-core/src/safe_client.rs [hunk 6]
```diff
@@ -47,6 +49,16 @@ impl<C> SafeClient<C> {
         &mut self.authority_client
     }
 
+    fn get_committee(&self, epoch_id: &EpochId) -> SuiResult<Committee> {
+        match self.epoch_store.get_authenticated_epoch(epoch_id)? {
```
Comment (gdanezis) [inline crates/sui-core/src/safe_client.rs:53]: Does that do a DB read? If so I would try to minimise its repeated use in the functions (call it once at the start).

File: crates/sui-types/src/messages_checkpoint.rs [hunk 7]
```diff
@@ -123,38 +123,6 @@ pub struct CheckpointResponse {
     pub detail: Option<CheckpointContents>,
 }
 
-impl CheckpointResponse {
```
Comment (gdanezis) [inline crates/sui-types/src/messages_checkpoint.rs:126]: Is this refactoring related to this PR and epoch DB?

**Review by gdanezis — APPROVED** [review]
Cool stuff -- have a look at the comments. In particular calling to get the latest committee once to avoid inconsistent reads?

File: crates/sui-core/src/authority.rs (hunk 2 above)
Comment (lxfind) [reply crates/sui-core/src/authority.rs:1051]: We don't have that support yet. We will need to pass some seed node addresses at start for the node to sync.

File: crates/sui-types/src/messages_checkpoint.rs (hunk 7 above)
Comment (lxfind) [reply crates/sui-types/src/messages_checkpoint.rs:126]: Yes I had to move this check into within SafeClient due to the use of the epoch db.
